### PR TITLE
Add new positive_decimal rule

### LINKED ISSUE #86: [2.01] Add rule: positive decimal in specific elements
The elements in the following list should enforce a positive decimal

```
iati-activities/iati-activity/recipient-country/@percentage
iati-activities/iati-activity/recipient-region/@percentage
iati-activities/iati-activity/sector/@percentage
iati-activities/iati-activity/capital-spend/@percentage
```

Add the relevant rule and tests

### PR DESCRIPTION
**Pull request by samuele-mattiuzzo**
Fixes #86

### CODE REVIEW

**Comment by samuele-mattiuzzo** [comment]
@Ocre42 this is still not complete, I have forgotten to update the php validator test case. Will do it later on this evening from home or tomorrow morning first thing when I come in

**Comment by samuele-mattiuzzo** [comment]
@Ocre42 I still need to do some work on this, I think I got it all wrong, so will move this back to in progress

File: rulesets/standard.json [hunk 1]
```diff
@@ -63,6 +63,11 @@
             ]
         }
     },
+    "positive_decimal": {
+        "cases": [ {
+            "paths": [ "recipient-country/@percentage", "recipient-region/@percentage", "sector/@percentage", "capital-spend/@percentage" ]
+         } ]
+    },
```
Comment (andylolz) [inline rulesets/standard.json]: This bit doesn’t look right to me – I think the curly bracket on L65 should be moved down a bit.

The `positive_decimal` rule is currently at the same level as `//iati-activity` and `//iati-organisation` (the `context_xpath`). `Rule`s should be one level down from these `context_xpath`s.

Comment (andylolz) [reply rulesets/standard.json]: I think it’s correct in meta_tests/positive_decimal.json, but wrong here.

**Comment by andylolz** [comment]
This is COOL – it is great that you’re pulling these rules together here.

I’ve added one comment, because while this PR looks mostly great, I think it might introduce a bug. The test case is fine (hence it passes) but the bug is in standard.json.

**Comment by andylolz** [comment]
> This positive_decimal rule will probably change into something more self contained that uses xml builtins to check both for positive and decimal

Yeah, sounds good. If I understand correctly, then maybe it’s worth taking a look at the following PR which I think follows the approach you describe:
https://github.com/IATI/IATI-Schemas/pull/329

**Comment by samuele-mattiuzzo** [comment]
Aaaah good find! I'll take a look at it, but yeah, Ideally having a single line that assesses that `elem.value >= 0.0` is what I'm aiming for (so `positive_decimal` is probably going to go away anyway) to allow validators to just evaluate the expression as a `true` / `false` (rather than doing what I'm doing in python: retrieve the value, convert to decimal, check it against `>=0.0`)

basically I want to make it stand alone and language agnostic and maybe more generic (something that adds the operator and the sentinel value as extra params to be checked/combined, although this wouldn't work with my above example)

**Comment by andylolz** [comment]
So I think you’re looking for a new `evaluates_to_true` rule (both here, and in the `if_then` thread). Then you’d do something like:

```json
{
    "//iati-activity": {
        "evaluates_to_true": {
            "cases": [{
                "paths": [
                    "number(recipient-country/@percentage) >= 0",
                    "number(recipient-region/@percentage) >= 0",
                    "number(sector/@percentage) >= 0",
                    "number(capital-spend/@percentage) >= 0"
                ]
            }]
        }
    }
}
```

But I think Vincent’s PR is even better, because a) it’s in XSD, and b) it checks upper and lower bounds (because of course for percentage, `100 >= elem.value >= 0.0`).

Enforcing rules at schema level is language agnostic & generic – my understanding is that rulesets should just be for stuff that is difficult or impossible to do in XSD.

**Comment by andylolz** [comment]
> although we're checking for a `positive_decimal` here, if we add a range as well `100.0 >= elem.value >= 0.0` it's semantically different

Yeah absolutely. All the examples in this case are `@percentage`, so that’s why I mention bounds / Vincent’s PR.

**Pull request merged**